# Задание 1

Для датафрейма log из материалов занятия создайте столбец source_type по следующим правилам:

- если источник traffic_source равен yandex или google, то в source_type ставится organic
- для источников paid и email из России - ставим ad
- для источников paid и email не из России - ставим other
- все остальные варианты берем из traffic_source без изменений

In [2]:
import pandas as pd

In [43]:
logfile = pd.read_csv('/Users/zorinapancenko/Desktop/python/Downloads/visit_log.csv', sep = ';')

In [356]:
def newcolumn(data, inplace = True):
    
# выскакивала ошибка unsupported operand type(s) for |: 'str' and 'str', поэтому перевела данные в целые числа при помощи int()

    if int(data['traffic_source'] == 'yandex') | int(data['traffic_source'] == 'google'):
        return 'organic'
    elif (int(data['traffic_source'] == 'paid') | int(data['traffic_source'] == 'email')) & int(data['region'] == 'Russia'):
        return 'ad'
    elif (int(data['traffic_source'] == 'paid') | int(data['traffic_source'] == 'email')) & int(data['region'] != 'Russia'):
        return 'other'
    else: 
        return data['traffic_source']

In [357]:
logfile['source_type'] = logfile.apply(newcolumn, axis=1)
logfile.head(20)

timestamp    visit_id                                 url   region  \
0   1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3   Russia   
1   1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a   Russia   
2   1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15   Russia   
3   1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15   Russia   
4   1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15   Russia   
5   1549980742  8855508aad  https://host.ru/df646c3676cc259fa0   Russia   
6   1549980742  b0f66adc83  https://host.ru/b8b58337d272ee7b15   Russia   
7   1549980754  837885c8f8  https://host.ru/108ce4b365afb7b88e   Russia   
8   1549980760  af5570f5a1  https://host.ru/3004a8273caeef2867    China   
9   1549980765  3e7077fd2f  https://host.ru/df646c3676cc259fa0   Russia   
10  1549980777  01d448afd9  https://host.ru/a5dda93e70318570c0    India   
11  1549980785  71b6c1d538  https://host.ru/c8d9213a31839f9a3a  Ukraine   
12  1549980789  15ec7bf0b5  https://host.ru/8e88d4703848cc0ec4   Russia   
13  1549980792  dd46c3eebb  https://host.ru/3004a8273caeef2867    China   
14  1549980793  e7ecebbc59  https://host.ru/05e41944dd85a900a6  Belarus   
15  1549980797  5322fecfc9  https://host.ru/4adc8a29b212051b04  Belarus   
16  1549980801  374708fff7  https://host.ru/39fa884393666d45fc   Russia   
17  1549980807  0a88111852  https://host.ru/2fcb223266880ccf57  Ukraine   
18  1549980814  60daa3a5f7  https://host.ru/ef57543002ce2a0f2a  Belarus   
19  1549980816  d6fd62f5ce  https://host.ru/f1eb4601740d627ab0  Belarus   

       user_id traffic_source source_type  
0   b1613cc09f         yandex     organic  
1   4c3ec14bee         direct      direct  
2   a8c40697fb         yandex     organic  
3   521ac1d6a0         yandex     organic  
4   d7323c571c         yandex     organic  
5   fc43898e47         yandex     organic  
6   13fc55e781           paid          ad  
7   cb5082b6f6         direct      direct  
8   45664f7af2         direct      direct  
9   6f9de8c8b6          email          ad  
10  27e76a4a80         direct      direct  
11  3b13701488         google     organic  
12  6d1d81f7f8           paid          ad  
13  7e301c8c41         direct      direct  
14  ac52dedc3b         direct      direct  
15  9a45d98867           paid       other  
16  e9bd59e22d         google     organic  
17  a2ba8fa3db           paid       other  
18  72cb10fb5c         direct      direct  
19  7269dcd869         direct      direct

In [358]:
logfile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18938 entries, 0 to 18937
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   timestamp       18938 non-null  int64 
 1   visit_id        18938 non-null  object
 2   url             18938 non-null  object
 3   region          18938 non-null  object
 4   user_id         18938 non-null  object
 5   traffic_source  18938 non-null  object
 6   source_type     18938 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


# Задание 2

В файле URLs.txt содержатся url страниц новостного сайта. Вам необходимо отфильтровать его по адресам страниц с текстами новостей. Известно, что шаблон страницы новостей имеет внутри url следующую конструкцию: /, затем 8 цифр, затем дефис. Выполните следующие действия:

- Прочитайте содержимое файла с датафрейм
- Отфильтруйте страницы с текстом новостей, используя метод str.contains и регулярное выражение в соответствии с заданным шаблоном

In [55]:
import re

In [69]:
urlfile = pd.read_csv('/Users/zorinapancenko/Desktop/python/Downloads/URLs.txt', sep = '\t')
urlfile.head()

url
0                                            /world/
1                                           /latest/
2                                      /?updated=top
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...

In [66]:
newfile = urlfile[ urlfile['url'].str.contains('/\d{8}-')]
newfile

url
3   /politics/36188461-s-marta-zhizn-rossiyan-susc...
4   /world/36007585-tramp-pridumal-kak-reshit-ukra...
5   /science/36157853-nasa-sobiraet-ekstrennuyu-pr...
6   /video/36001498-poyavilis-pervye-podrobnosti-g...
7   /world/36007585-tramp-pridumal-kak-reshit-ukra...
..                                                ...
88  /cis/35984145-kreml-prokommentiroval-soobschen...
89  /video/36071019-olimpiyskie-obekty-rio-prevrat...
90  /science/36151301-nazvano-posledstvie-zloupotr...
91  /incidents/36027330-vospitatelnitsu-zatravili-...
92  /world/36103095-dominikanskih-zhurnalistov-ubi...

[84 rows x 1 columns]

In [67]:
len(urlfile)

93

In [68]:
len(newfile)

84

# Задание 3

Используйте файл с оценками фильмов ml-latest-small/ratings.csv. Посчитайте среднее время жизни пользователей, которые выставили более 100 оценок. Под временем жизни понимается разница между максимальным и минимальным значением столбца timestamp для данного значения userId.

In [70]:
ratings = pd.read_csv('/Users/zorinapancenko/Desktop/python/Downloads/ratings.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [98]:
ratings_func = ratings.groupby('userId').agg([min, max])
ratings_func.head()

movieId        rating        timestamp            
           min    max    min  max         min         max
userId                                                   
1           31   3671    1.0  4.0  1260759108  1260759205
2           10    720    1.0  5.0   835355395   835356246
3           60  84236    2.0  5.0  1298861589  1298932787
4           10   4006    1.0  5.0   949778714   949982274
5            3  48385    1.5  5.0  1163373044  1163375145

In [149]:
import warnings

warnings.filterwarnings('ignore')

In [366]:
# создаю отдельный датафрейм movie_fans с числом оценок пользователей (по условию number >100) и затем объединяю его с датафреймом ratings_func (к которому применила ф-ции max и min)
ratings_count = ratings.groupby('userId').count()[['rating']]
movie_fans = ratings_count[ratings_count['rating']>100]
# переименую столбец с числом оценок перед объединением таблиц
movie_fans.rename(columns = {'rating': 'number'}, inplace=True)

joined = movie_fans.merge(ratings_func, how = 'left', on = 'userId')

In [367]:
joined['lifetime'] = joined[('timestamp', 'max')] - joined[('timestamp', 'min')]
joined.head(10)

number  (movieId, min)  (movieId, max)  (rating, min)  (rating, max)  \
userId                                                                         
4          204              10            4006            1.0            5.0   
8          116              32           44004            0.5            5.0   
15        1700               1          161155            0.5            5.0   
17         363               6           34437            0.5            5.0   
19         423               1            4970            1.0            5.0   
21         162              10            5060            2.0            5.0   
22         220              32           37729            1.0            5.0   
23         726               1           48385            0.5            5.0   
26         172               1           92259            0.5            5.0   
30        1011               1            6565            1.0            5.0   

        (timestamp, min)  (timestamp, max)   lifetime  
userId                                                 
4              949778714         949982274     203560  
8             1154389340        1154474527      85187  
15             997937239        1469330735  471393496  
17            1127468587        1127476640       8053  
19             855190091         855195373       5282  
21             853157476         854522908    1365432  
22            1131661890        1131753381      91491  
23            1148386124        1166728253   18342129  
26            1351544316        1371811577   20267261  
30             944943070        1060795346  115852276

In [364]:
print(joined['lifetime'].mean())

joined['lifetime'].mean()/86400

40080507.4496124


463.8947621482917

### Ответ: среднее время жизни пользователей, которые выставили более 100 оценок, приблизительно равно 40080507 в единицах измерения Unix time или 464 дня

# Задание 4

Дана статистика услуг перевозок клиентов компании по типам (см. файл “Python_13_join.ipynb” в разделе Материалы для лекции «Продвинутый pandas» ---- Ноутбуки к лекции «Продвинутый pandas»).
Необходимо сформировать две таблицы:
- таблицу с тремя типами выручки для каждого client_id без указания адреса клиента
- аналогичную таблицу по типам выручки с указанием адреса клиента

Обратите внимание, что в процессе объединения таблиц данные не должны теряться.

In [246]:
rzd = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115],
        'rzd_revenue': [1093, 2810, 10283, 5774, 981]
    }
)

auto = pd.DataFrame(
    {
        'client_id': [113, 114, 115, 116, 117],
        'auto_revenue': [57483, 83, 912, 4834, 98]
    }
)

air = pd.DataFrame(
    {
        'client_id': [115, 116, 117, 118],
        'air_revenue': [81, 4, 13, 173]
    }
)

client_base = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115, 116, 117, 118],
        'address': ['Комсомольская 4', 'Энтузиастов 8а', 'Левобережная 1а', 'Мира 14', 'ЗЖБИиДК 1', 
                    'Строителей 18', 'Панфиловская 33', 'Мастеркова 4']
    }
)


In [328]:
newtable_1 = rzd.merge(auto, on='client_id', how = 'outer').merge(air, on='client_id', how = 'outer')
newtable_1

client_id  rzd_revenue  auto_revenue  air_revenue
0        111       1093.0           NaN          NaN
1        112       2810.0           NaN          NaN
2        113      10283.0       57483.0          NaN
3        114       5774.0          83.0          NaN
4        115        981.0         912.0         81.0
5        116          NaN        4834.0          4.0
6        117          NaN          98.0         13.0
7        118          NaN           NaN        173.0

In [331]:
newtable_2 = rzd.merge(auto, on='client_id', how = 'outer').merge(air, on='client_id', how = 'outer').merge(client_base, on='client_id', how = 'outer')
newtable_2


client_id  rzd_revenue  auto_revenue  air_revenue          address
0        111       1093.0           NaN          NaN  Комсомольская 4
1        112       2810.0           NaN          NaN   Энтузиастов 8а
2        113      10283.0       57483.0          NaN  Левобережная 1а
3        114       5774.0          83.0          NaN          Мира 14
4        115        981.0         912.0         81.0        ЗЖБИиДК 1
5        116          NaN        4834.0          4.0    Строителей 18
6        117          NaN          98.0         13.0  Панфиловская 33
7        118          NaN           NaN        173.0     Мастеркова 4